test/helpers/envtest.go

```diff
@@ -130,3 +134,35 @@ func (t *TestEnvironment) Stop() error {
 func (t *TestEnvironment) CreateKubeconfigSecret(cluster *clusterv1.Cluster) error {
 	return kubeconfig.CreateEnvTestSecret(t.Client, t.Config, cluster)
 }
+
+type cleanup func() error
```

Comment:
I noticed that in the [kubebuilder envtest docs](https://book.kubebuilder.io/reference/testing/envtest.html#testing-considerations)
> One common example is garbage collection; because there are no controllers monitoring built-in resources, objects do not get deleted, even if an OwnerReference is set up.

That is, just deleting the namespace will not cascade deletes to the individual objects.  
I added some helpers like `CreateNamespace` and `CreateObj` which returns a `cleanup` function so it forces the user to handle the cleanup of these objects vs. providing a separate delete helper.

Other thoughts:
I know that deleting objects may not be an issue if everything is properly namespaced but I think there are some common objects like `kubeadm-config` in common namespaces like `kube-system` or even cluster wide resources that may still need to be properly cleaned up so as to avoid cross-test contamination.

Reply:
This is something we should document in the testing guidelines.
WRT to the solution, I'm personally in favor of enforcing a pattern where we should have test/test cases to run in random namespaces/use random object names instead of relying on cleanup, mainly because this simplifies writing tests (no need of the cleanup logic).

A possible middle ground is to have a delete ns func that enforce deletion of all the objects in a namespace, but if possible I would avoid requiring developers to take care of deletion of single objects.

Reply:
I guess we can document the cleanup as an _option_ if the developer needs to do it. That is, they can ignore the `cleanup` function returned. `_, err := testEnv.CreateObj(ctx, o)`